### PR TITLE
opal/mca/accelerator: Changing component selection logic and adding NULL Component

### PR DESCRIPTION
**Pull request by a-szegel**
Changing component selection logic from priority based system, to a system that assumes that there will only ever be 1 type of accelerator on an instance.  Added a Null Accelerator Component so that accelerator selection logic never fails.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

### CODE REVIEW

**Comment by a-szegel** [comment]
I went ahead and removed all the checks in the second commit that would impact performance.  I don't have confidence in these changes b/c I am not sure how to test them, and I am not sure if my changes are correct (which is why I didn't rebase into 1 commit).

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+static void accelerator_null_finalize(void)
+{
+    return;
+}
+
+/* Accelerator API's Implementation */
+static int accelerator_null_check_addr(const void *addr, int *dev_id, uint64_t *flags)
+{
+    opal_output(0, "Error: Called method on accelerator_null_check_addr() on Null Accelerator");
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_null_component.c:150]: I don't think this is right; This is the function that should work and just always return `0` (not accelerator memory).  Otherwise, what is the point of this component?  Every caller would have to have error handling in the critical path, which is not an improvement over #ifs.

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    return -1;
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_null_component.c:151]: Return codes should always be OPAL error codes (ie, `OPAL_ERROR` or `OPAL_ERR_NOT_IMPLEMENTED`).

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+static int accelerator_null_memcpy(int dev_id, void *dest, const void *src,
+                            size_t size, opal_accelerator_transfer_type_t type)
+{
+    opal_output(0, "Error: Called method on accelerator_null_memcpy() on Null Accelerator");
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_null_component.c:157]: Why should we print output here?  If we're willing to print output, should we also abort?

File: opal/mca/accelerator/base/accelerator_base_null_component.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#ifndef MCA_ACCELERATOR_BASE_NULL_COMPONENT_H
+#define MCA_ACCELERATOR_BASE_NULL_COMPONENT_H
+
+#include "opal_config.h"
+#include "opal/mca/mca.h"
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_null_component.h:15]: In OMPI, our style is to have minimal includes in the header files, and the source files include the headers they need.  We do not use this style of a full list of headers in the main header.  Also, many of these headers look unnecessary for this component; please prune.

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 5]
```diff
@@ -63,83 +52,32 @@ int opal_accelerator_base_select(void)
...
+        module = component->accelerator_init();
+        if (NULL != module && !found_component) {
+            found_component = true;
+            accelerator_base_selected_component = *component;
+            opal_accelerator = *module;
         }
-        OBJ_DESTRUCT(aoc);
-        free(aoc);
+        else if (NULL != module && found_component) {
+            opal_output(0,
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_select.c:67]: This should be a show-help

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 6]
```diff
@@ -63,83 +52,32 @@ int opal_accelerator_base_select(void)
...
-                   recover, anyway?  This component is going away, so errors
-                   don't matter anymore) */
-
-                aoc->accelerator_component->accelerator_finalize();
-                opal_output_verbose(10, opal_accelerator_base_framework.framework_output,
-                                    "select: component %s not selected / finalized",
-                                    aoc->accelerator_component->base_version.mca_component_name);
-            }
+        module = component->accelerator_init();
+        if (NULL != module && !found_component) {
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_select.c:61]: This seems overly complicated, given you're going to essentially abort in one of the two cases.  Why not just have
```
if (NULL != module) {
   if (found_component) {
    opal_show_help(...);
    return NULL;
  }
  found_component = true;
   ....
```

Much cleaner that way.  Also, OMPI style is generally to put the end bracket and the else clause on the same line, so:

```
  } else if (...) {
```
not
```
  }
  else if (...) {
```

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 7]
```diff
@@ -63,83 +52,32 @@ int opal_accelerator_base_select(void)
...
-    OBJ_DESTRUCT(&opened);
 
-    /* This base function closes, unloads, and removes from the
-       available list all unselected components.  The available list will
-       contain only the selected component. */
-
-    mca_base_components_close(opal_accelerator_base_framework.framework_output,
-                              &opal_accelerator_base_framework.framework_components,
-                              (mca_base_component_t *) best_component);
+    if (NULL == module) {
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/accelerator_base_select.c:76]: you shouldn't need this.  In fact, it's wrong.  The NULL component is just another component; it's in the list above.  Thinking about it more (sorry, I missed this when we were talking yesterday), if you're going to have a NULL component, you need to use priorities and handle selecting the first one that succeeds and ordering and all that (which kind of sucks), because otherwise you'll always have two components if there is an accelerator (the one that works and the one that is the NULL component).  You could probably explicitly check for that case instead (and skip the NULL component initialization), but you somehowe need to handle the fact that the NULL component is in the list above.

File: opal/mca/accelerator/base/base.h [hunk 8]
```diff
@@ -27,9 +27,9 @@ BEGIN_C_DECLS
 OPAL_DECLSPEC extern mca_base_framework_t opal_accelerator_base_framework;
 
 /**
- * Select an accelerator module
+ * Select an ac
```
Comment (bwbarrett) [inline opal/mca/accelerator/base/base.h:30]: ?

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 7 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:76]: My thought was that I didn't want to register the NULL component so that this logic would be correct.

File: opal/mca/accelerator/base/base.h (hunk 8 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/base.h:30]: fixed.

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 9]
```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (c) 2017-2022 Amazon.com, Inc. or its affiliates.
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:2]: I copied accelerator_cuda.c, and deleted all the impl.  I will pull the rest of the copyrights in here.

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+ *                         All Rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ *
+ * These symbols are in a file by themselves to provide nice linker
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:10]: removed, I copied both accelerator_cuda_component.c, and accelerator_cuda.c into one file (made sure to include all names on accelerator_base_null_component files).

[the remark this answers is not in the document]
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:10]: I copied the header from accelerator_cuda.h, so I just took the header from that.

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+ * semantics.  Since linkers generally pull in symbols by object
+ * files, keeping these symbols as the only symbols in this file
+ * prevents utility programs such as "ompi_info" from having to import
+ * entire components just to query their version and parameters.
+ */
+
+#include "accelerator_base_null_component.h"
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:17]: done

File: opal/mca/accelerator/base/accelerator_base_null_component.c [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+{
+    return OPAL_SUCCESS;
+}
+
+static opal_accelerator_base_module_t* accelerator_null_init(void)
+{
+    return &opal_accelerator_null_module;
+}
+
+static void accelerator_null_finalize(void)
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:142]: Done, updated the Accelerator API, CUDA, and NULL Components

File: opal/mca/accelerator/base/accelerator_base_null_component.c (hunk 2 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:151]: Switched all RV's to OPAL_ERR_NOT_IMPLEMENTED.

File: opal/mca/accelerator/base/accelerator_base_null_component.c (hunk 1 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.c:150]: yes, we talked about this and I forgot.  Updated.

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 7 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:76]: just confirmed... the implementation only goes around the OPAL_LIST_FOREACH once, the NULL Component isn't registered in ` &opal_accelerator_base_framework.framework_components` variable

Comment (bwbarrett) [reply opal/mca/accelerator/base/accelerator_base_select.c:76]: then something is broken; it certainly should be in there.

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 6 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:61]: done

File: opal/mca/accelerator/base/accelerator_base_null_component.h (hunk 4 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_null_component.h:15]: pruned

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 5 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:67]: switched to show help

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 7 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:76]: broken as intended :)... will investigate.

**Comment by a-szegel** [comment]
Remaining actions:
1. Figure out why NULL Component isn't being registered as a component
2. Figure out what NULL Components return codes should be for methods that shouldn't be called (OPAL_SUCCESS, OPAL_NOT_IMPLEMENTED, or an abort)

File: ompi/mca/common/ompio/common_ompio_buffer.c [hunk 13]
```diff
@@ -144,7 +148,7 @@ void *mca_common_ompio_alloc_buf ( ompio_file_t *fh, size_t bufsize )
     if ( !mca_common_ompio_buffer_init ){
         mca_common_ompio_buffer_alloc_init ();
     }
-
```
Comment (wckzhang) [inline ompi/mca/common/ompio/common_ompio_buffer.c:147]: Random whitespace changes throughout this commit, can you split whitespace changes to a separate commit?

File: ompi/mca/common/ompio/common_ompio_buffer.c [hunk 14]
```diff
@@ -82,8 +83,11 @@ static void* mca_common_ompio_buffer_alloc_seg ( void*ctx, size_t *size )
 
 static void mca_common_ompio_buffer_free_seg ( void *ctx, void *buf )
 {
+    int tmp = -1;
+    uint64_t flags = -1;
+
     if ( NULL != buf ) {
-        if (NULL != opal_accelerator.host_unregister) {
+        if (opal_accelerator.check_addr(buf, &tmp, &flags) > 0) {
```
Comment (wckzhang) [inline ompi/mca/common/ompio/common_ompio_buffer.c:90]: I think this check is wrong. buf is supposed to be a host buffer, not a device buffer. Read here:

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ge8d5c17670f16ac4fc8fcb4181cb490c

File: ompi/mca/common/ompio/common_ompio_buffer.c [hunk 15]
```diff
@@ -66,14 +63,18 @@ static void* mca_common_ompio_buffer_alloc_seg ( void*ctx, size_t *size )
...
 
     numpages = (*size + mca_common_ompio_pagesize -1 )/mca_common_ompio_pagesize;
     realsize = numpages * mca_common_ompio_pagesize;
 
     buf = malloc ( realsize);
 
-    if (NULL != buf && NULL != opal_accelerator.host_register) {
-        opal_accelerator.host_register((void *)buf, realsize);
+    if (NULL != buf) {
+        if (opal_accelerator.check_addr(buf, &tmp, &flags) > 0) {
```
Comment (wckzhang) [inline ompi/mca/common/ompio/common_ompio_buffer.c:75]: Pretty sure this check is wrong, see:

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ge8d5c17670f16ac4fc8fcb4181cb490c

File: opal/datatype/opal_convertor.c [hunk 16]
```diff
@@ -45,11 +45,14 @@
...
 {
     int res, dev_id;
+    int tmp = -1;
+    uint64_t flags = 0;
     if (!(convertor->flags & CONVERTOR_ACCELERATOR)) {
         return MEMCPY(dest, src, size);
     }
 
-    if (NULL != opal_accelerator.memcpy) {
+    if ((opal_accelerator.check_addr(dest, &tmp, &flags) > 0) ||
```
Comment (wckzhang) [inline opal/datatype/opal_convertor.c:54]: If we're going to use these checks anyway, it's probably worth setting the DtoH or DtoD, etc. flags. The current cuda impl doesn't utilize them, but if we're already doing these checks, it makes sense to pass the information through. Same comment for similar checks

File: opal/mca/accelerator/accelerator.h [hunk 17]
```diff
@@ -308,7 +308,7 @@ typedef struct {
  *
  * @return                   Initialized module or NULL if init failed.
  */
-typedef opal_accelerator_base_module_t * (*mca_accelerator_base_component_init_fn_t)();
+typedef opal_accelerator_base_module_t * (*mca_accelerator_base_component_init_fn_t)(void);
```
Comment (wckzhang) [inline opal/mca/accelerator/accelerator.h:311]: This should be part of another commit

**Review by wckzhang — CHANGES_REQUESTED** [review]

File: opal/mca/accelerator/accelerator.h [hunk 18]
```diff
@@ -323,9 +323,11 @@ typedef opal_accelerator_base_module_t * (*mca_accelerator_base_component_init_f
  * during the lifetime of the component, including outstanding
  * modules.
  *
+ * @param  module     Allow the component to perform cleanup in module
```
Comment (wckzhang) [inline opal/mca/accelerator/accelerator.h:323]: The style is different from the rest of the comments. I'm also not clear if this means the component is supposed to free the module or not

File: opal/mca/accelerator/null/accelerator_null_component.c [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+        },
+    /* Next the MCA v1.0.0 component meta data */
+    .base_data =
+        {/* The component is checkpoint ready */
+         MCA_BASE_METADATA_PARAM_CHECKPOINT},
+    .accelerator_init = accelerator_null_init,
+    .accelerator_finalize = accelerator_null_finalize,
+}};
+
+opal_accelerator_base_module_t opal_accelerator_null_module =
```
Comment (wckzhang) [inline opal/mca/accelerator/null/accelerator_null_component.c:102]: Can you update with my asynchronous API's

File: opal/mca/accelerator/null/accelerator_null_component.c [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+
+static void accelerator_null_finalize(opal_accelerator_base_module_t* module)
+{
+    return;
+}
+
+/* Accelerator API's Implementation */
+static int accelerator_null_check_addr(const void *addr, int *dev_id, uint64_t *flags)
+{
+    /* Always return that the pointer belongs to the accelerator */
```
Comment (wckzhang) [inline opal/mca/accelerator/null/accelerator_null_component.c:153]: This comment is incorrect

File: opal/mca/accelerator/cuda/accelerator_cuda_component.c [hunk 21]
```diff
@@ -34,7 +34,7 @@ static int accelerator_cuda_open(void);
 static int accelerator_cuda_close(void);
 static int accelerator_cuda_component_register(void);
 static opal_accelerator_base_module_t* accelerator_cuda_init(int *priority);
-static void accelerator_cuda_finalize(void);
```
Comment (wckzhang) [inline opal/mca/accelerator/cuda/accelerator_cuda_component.c:37]: Overall commit feedback, null component should be a separate commit from changing selection logic. Would make it much easier for me to append to my commits.

File: opal/mca/accelerator/accelerator.h [hunk 22]
```diff
@@ -306,12 +306,9 @@ typedef struct {
  * We assume that only one accelerator will be present
  * on any given node.
  *
- * @param[OUT]  priority     Relative priority or ranking used by MCA to
- *                           select a component.
  * @return                   Initialized module or NULL if init failed.
  */
-typedef opal_accelerator_base_module_t * (*mca_accelerator_base_component_init_fn_t)(
-    int *priority);
```
Comment (wckzhang) [inline opal/mca/accelerator/accelerator.h:314]: So reading over the selection logic, I think it's more confusing than what was previously there. There is an unnecessary amount of list manipulation imo. Why not just use the priority code that existed and return a fatal error if the initialized list length > 2?

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 23]
```diff
@@ -32,116 +32,108 @@
 #include "opal/mca/mca.h"
 #include "opal/mca/accelerator/base/base.h"
 #include "opal/mca/accelerator/accelerator.h"
+#include <string.h>
 
-typedef struct accelerator_opened_component_t {
+typedef struct accelerator_component_t {
   opal_list_item_t super;
   opal_accelerator_base_component_t *accelerator_component;
-  opal_accelerator_base_module_t *accelerator_module;
```
Comment (wckzhang) [inline opal/mca/accelerator/base/accelerator_base_select.c]: Really confused, didn't you just add this in the last commit? Why does one PR have a change that was added and removed?

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 24]
```diff
@@ -32,116 +32,108 @@
...
+
+        if (0 == strcmp(cli->cli_component->mca_component_name, "null")) {
+            opal_list_append(&ordered_list, (opal_list_item_t*) ac);
+        }
+        else {
+            opal_list_prepend(&ordered_list, (opal_list_item_t*) ac);
+        }
+    }
+
+    /* Traverse the ordered list of available components;
```
Comment (wckzhang) [inline opal/mca/accelerator/base/accelerator_base_select.c:73]: I don't think this style of multi-line commenting matches the rest of OMPI code base

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 25]
```diff
@@ -32,116 +32,108 @@
...
-
-                aoc->accelerator_component->accelerator_finalize(aoc->accelerator_module);
-                opal_output_verbose(10, opal_accelerator_base_framework.framework_output,
-                                    "select: component %s not selected / finalized",
-                                    aoc->accelerator_component->base_version.mca_component_name);
+        module = component->accelerator_init();
+        if (NULL != module) {
+            /* A second non NULL component (the final entry in the list) has initialized successfully */
+            if (found_component) {
+                if (0 != ordered_list.opal_list_length) {
```
Comment (wckzhang) [inline opal/mca/accelerator/base/accelerator_base_select.c:96]: I don't know too well, but I don't think this check will ever be true?

File: opal/mca/accelerator/base/accelerator_base_select.c [hunk 26]
```diff
@@ -32,116 +32,108 @@
...
+                    accelerator_base_selected_component.accelerator_finalize(&opal_accelerator);
+                    component->accelerator_finalize(module);
+
+                    /* Throw a fatal error, can not go on without knowing which Accelerator to use
+                    Leak some memory but doesn't matter b/c OMPI is aborting */
+                    return OPAL_ERR_FATAL;
+                }
+
+                /* Prefer found component over NULL component*/
+                // OBJ_DESTRUCT(item);
```
Comment (wckzhang) [inline opal/mca/accelerator/base/accelerator_base_select.c:111]: Leftover code?

File: opal/mca/accelerator/null/accelerator_null_component.c [hunk 27]
```diff
@@ -31,7 +31,7 @@ const char *opal_accelerator_null_component_version_string
 static int accelerator_null_open(void);
 static int accelerator_null_close(void);
 static int accelerator_null_component_register(void);
-static opal_accelerator_base_module_t* accelerator_null_init(int *priority);
```
Comment (wckzhang) [inline opal/mca/accelerator/null/accelerator_null_component.c]: This commit should have occurred before the last one, no reason to add something and then change it immediately in the next commit

**Review by wckzhang — CHANGES_REQUESTED** [review]

File: opal/mca/accelerator/accelerator.h (hunk 18 above)
Comment (a-szegel) [reply opal/mca/accelerator/accelerator.h:323]: If the component performed allocation in the module class, we give it the option to clean that allocation up.

Comment (a-szegel) [reply opal/mca/accelerator/accelerator.h:323]: Updated comment description to be more clear, and added [in] to match style

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 25 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:96]: Yes, this is saying:
```
If a component initializes successfully
   If We have Already Found our Component
       If the component is not the NULL component
```

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 26 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:111]: no, I actually want this.  Commented it out when I was debugging a seg fault, but forgot to put this back in.

File: opal/mca/accelerator/null/accelerator_null_component.c (hunk 27 above)
Comment (a-szegel) [reply opal/mca/accelerator/null/accelerator_null_component.c]: Moved the component selection commit before the NULL Accelerator Commit

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 25 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:96]: changed the check to be easier to read

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 24 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:73]: Updated style to match other multi line comments

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 23 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c]: I think this change is gone

File: opal/mca/accelerator/accelerator.h (hunk 22 above)
Comment (a-szegel) [reply opal/mca/accelerator/accelerator.h:314]: discussed offline, fine as is

File: opal/mca/accelerator/cuda/accelerator_cuda_component.c (hunk 21 above)
Comment (a-szegel) [reply opal/mca/accelerator/cuda/accelerator_cuda_component.c:37]: done

File: opal/mca/accelerator/null/accelerator_null_component.c (hunk 20 above)
Comment (a-szegel) [reply opal/mca/accelerator/null/accelerator_null_component.c:153]: corrected accelerator -> host

File: opal/mca/accelerator/accelerator.h (hunk 17 above)
Comment (a-szegel) [reply opal/mca/accelerator/accelerator.h:442]: done

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 25 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c:94]: my attempt to make check better failed, reverted to old check but added comments.

File: opal/mca/accelerator/base/accelerator_base_select.c (hunk 23 above)
Comment (a-szegel) [reply opal/mca/accelerator/base/accelerator_base_select.c]: gone b/c of re-ordering

File: opal/mca/accelerator/null/accelerator_null_component.c (hunk 19 above)
Comment (a-szegel) [reply opal/mca/accelerator/null/accelerator_null_component.c:114]: done

File: opal/datatype/opal_datatype_unpack.c [hunk 28]
```diff
@@ -184,6 +184,8 @@ opal_unpack_partial_predefined(opal_convertor_t *pConvertor, const dt_elem_desc_
     unsigned char *user_data = *memory + pElem->elem.disp;
     size_t data_length = opal_datatype_basicDatatypes[pElem->elem.common.type]->size;
     unsigned char *partial_data = *packed;
+    int tmp = -1;
+    uint64_t flags = -1;
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply opal/datatype/opal_datatype_unpack.c]: removed commit

File: opal/datatype/opal_convertor.c (hunk 16 above)
Comment (a-szegel) [reply opal/datatype/opal_convertor.c]: removed commit

File: ompi/mca/common/ompio/common_ompio_buffer.c (hunk 15 above)
Comment (a-szegel) [reply ompi/mca/common/ompio/common_ompio_buffer.c]: removed commit

File: ompi/mca/common/ompio/common_ompio_buffer.c (hunk 14 above)
Comment (a-szegel) [reply ompi/mca/common/ompio/common_ompio_buffer.c]: removed commit

File: ompi/mca/mtl/ofi/mtl_ofi.h [hunk 29]
```diff
@@ -363,14 +363,16 @@ ompi_mtl_ofi_deregister_buffer(ompi_mtl_ofi_request_t *ofi_req) {
...
+    uint64_t flags = 0;
     ret = ompi_mtl_ofi_deregister_buffer(ofi_req);
     if (OPAL_UNLIKELY(OMPI_SUCCESS != ret)) {
         return ret;
     }
     if (OPAL_UNLIKELY(NULL != ofi_req->buffer)) {
         if (!(ofi_req->convertor->flags & CONVERTOR_ACCELERATOR)) {
             free(ofi_req->buffer);
-        } else if (opal_accelerator.free != NULL) {
+        } else if (opal_accelerator.check_addr(buf, &tmp, &flags) > 0) {
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply ompi/mca/mtl/ofi/mtl_ofi.h]: removed commit

File: ompi/mca/common/ompio/common_ompio_buffer.c (hunk 13 above)
Comment (a-szegel) [reply ompi/mca/common/ompio/common_ompio_buffer.c]: removed commit

File: opal/mca/accelerator/accelerator.h (hunk 17 above)
Comment (wckzhang) [reply opal/mca/accelerator/accelerator.h:442]: I think you misunderstood what I meant. I meant this should be part of https://github.com/wckzhang/ompi/pull/3/commits/35e50fe19cd101d30a58c3664cd3b4d1f4ce947a which is where you removed int *priority

**Review by wckzhang — APPROVED** [review]
